### PR TITLE
Fixed #30099 -- Fixed invalid SQL when filtering a Subquery by an aggregate.

### PR DESCRIPTION
**Pull request by nasirhjafri**
fixed group by error in Filter by Count annotated inside Subquery

Fix for https://code.djangoproject.com/ticket/30099

### CODE REVIEW

File: tests/expressions/tests.py [hunk 1]
```diff
@@ -534,6 +534,33 @@ def test_subquery_references_joined_table_twice(self):
...
+            .values('cnt')[:1],
+            output_field=models.IntegerField()
+        )).values('float')
+
+        results = list(qs)
+        self.assertEqual(results, [
+            {
+                'float': 1.2
+            }
+        ])
```
Comment (charettes) [inline tests/expressions/tests.py]: Could do `self.assertEqual(qs.get()['float'], 1.2)`

File: tests/expressions/tests.py [hunk 2]
```diff
@@ -534,6 +534,33 @@ def test_subquery_references_joined_table_twice(self):
...
+            Employee(firstname="D", lastname="W", salary=1500),
+        ])
+
+        qs = Number.objects.annotate(min_valuable_count=Subquery(
+            Employee.objects.filter(
+                salary=OuterRef('integer')
+            )
+            .annotate(cnt=Count('salary'))
+            .filter(cnt__gt=2)
+            .order_by('cnt')
```
Comment (charettes) [inline tests/expressions/tests.py]: Not necessary AFAIK?

File: tests/expressions/tests.py [hunk 3]
```diff
@@ -534,6 +534,33 @@ def test_subquery_references_joined_table_twice(self):
...
+            .values('cnt')[:1],
+            output_field=models.IntegerField()
```
Comment (charettes) [inline tests/expressions/tests.py]: Not necessary AFAIK; `values('cnt')` infers it.

File: tests/expressions/tests.py [hunk 4]
```diff
@@ -534,6 +534,33 @@ def test_subquery_references_joined_table_twice(self):
         outer = Company.objects.filter(pk__in=Subquery(inner.values('pk')))
         self.assertFalse(outer.exists())
 
+    def test_filter_by_count_annotated_inside_subquery(self):
+        Number.objects.create(integer=1000, float=1.2)
+        Employee.objects.bulk_create([
+            Employee(firstname="A", lastname="Z", salary=1000),
+            Employee(firstname="B", lastname="Y", salary=1000),
+            Employee(firstname="C", lastname="X", salary=1000),
+            Employee(firstname="D", lastname="W", salary=1500),
```
Comment (charettes) [inline tests/expressions/tests.py]: You can drop all the `firstname`, `lastname` above and default to `''`. Also could create a single `Employee` and use `cnt__gt=0` below.

File: tests/expressions/tests.py [hunk 5]
```diff
@@ -534,6 +534,33 @@ def test_subquery_references_joined_table_twice(self):
         outer = Company.objects.filter(pk__in=Subquery(inner.values('pk')))
         self.assertFalse(outer.exists())
 
+    def test_filter_by_count_annotated_inside_subquery(self):
```
Comment (charettes) [inline tests/expressions/tests.py]: Let's make the test name aggregate function agnostic, what about the following?

```suggestion
    def test_subquery_filter_by_aggregate(self):
```

**Comment by nasirhjafri** [comment]
Hey @charettes Thanks for reviewing. I've updated the test cases accordingly. Can you please review it again?
Thanks,

File: tests/expressions/tests.py [hunk 6]
```diff
@@ -534,6 +534,21 @@ def test_subquery_references_joined_table_twice(self):
         outer = Company.objects.filter(pk__in=Subquery(inner.values('pk')))
         self.assertFalse(outer.exists())
 
+    def test_subquery_filter_by_aggregate(self):
+        Number.objects.create(integer=1000, float=1.2)
+        Employee.objects.create(firstname="", lastname="", salary=1000)
```
Comment (charettes) [inline tests/expressions/tests.py]: I should have been clearer but `firstname` and `lastname` can be omited as they'll default to `''` if missing.

**Comment by charettes** [comment]
buildbot, test on oracle.

**Pull request merged**